crates/knit-wit/src/lib.rs

```diff
@@ -0,0 +1,171 @@
+use anyhow::anyhow;
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+};
+use wit_component::{DecodedWasm, WitPrinter};
+use wit_parser::{PackageId, Resolve};
+
+wit_bindgen::generate!({
+    world: "knit-wit",
+});
+
+struct KnitWit;
+
+impl Guest for KnitWit {
+    fn knit_wit(
+        wit_paths: Vec<String>,
+        worlds: Vec<String>,
+        output_world: Option<String>,
+        output_package: Option<String>,
+        output_dir: Option<String>,
+    ) -> Result<Vec<(String, String)>, String> {
+        if wit_paths.is_empty() {
+            return Err("No wit-paths provided".to_owned());
+        }
+        if worlds.is_empty() {
+            return Err("No worlds provided to be knit".to_owned());
+        }
+
+        let output_package = output_package.unwrap_or("local:combined-wit".to_owned());
+        let output_world = output_world.unwrap_or("combined".to_owned());
+        let output_dir = output_dir.unwrap_or("combined-wit".to_owned());
+
+        let mut resolve = Resolve::default();
+        let id = resolve
+            .push_str(
+                "component.wit",
+                &target_wit_source(&output_package, &output_world),
+            )
+            .map_err(|e| e.to_string())?;
+
+        merge_wits(&mut resolve, wit_paths).map_err(|e| e.to_string())?;
+
+        let (base_world, _) = resolve
+            .worlds
+            .iter()
+            .find(|(world, _)| resolve.worlds[*world].name == output_world)
+            .expect("could not find base world");
+
+        merge_worlds(&mut resolve, base_world, worlds).map_err(|e| e.to_string())?;
+
+        resolve_to_readable_wit(resolve, id, output_dir).map_err(|e| e.to_string())
+    }
+}
+
+fn merge_wits(resolve: &mut Resolve, wit_paths: Vec<String>) -> anyhow::Result<()> {
+    for path in wit_paths {
+        let (temp, _) = parse_wit(&PathBuf::from(path)).map_err(|e| anyhow!("{:?}", e))?;
+        resolve.merge(temp).expect("could not merge wits");
+    }
+
+    Ok(())
+}
+
+fn merge_worlds(
+    resolve: &mut Resolve,
+    base_world: id_arena::Id<wit_parser::World>,
+    worlds: Vec<String>,
+) -> anyhow::Result<()> {
+    for world_name in worlds {
+        let (world_to_merge, _) = resolve
+            .worlds
+            .iter()
+            .find(|(world, _)| resolve.worlds[*world].name == world_name)
+            .unwrap();
+        resolve
+            .merge_worlds(world_to_merge, base_world)
+            .map_err(|e| {
+                anyhow!(
+                    "unable to merge with world '{}' due to: {}",
+                    resolve.worlds[world_to_merge].name,
+                    e
+                )
+            })?;
+    }
+
+    Ok(())
+}
+
+fn resolve_to_readable_wit(
```

Comment:
This code could be simplified quite a bit if you were to output a single file instead of a whole directory tree of files, e.g. https://github.com/bytecodealliance/wit-bindgen/blob/fa19e08a884ec62f95191319d8d296874424c736/crates/csharp/src/lib.rs#L766-L776

Reply:
That is the goal eventually but `ComponentizeJS` currently is on a much older `wit-parser` which does not support it.